src/propertySelector/propertySelector.js

```diff
@@ -0,0 +1,150 @@
+import checkBoxTpl from 'ui/dialog/tpl/checkbox';
+import buttonFactory from 'ui/button';
+import 'ui/propertySelector/css/propertySelector.css';
+import $ from 'jquery';
+
+export default function propertySelectorFactory(config) {
+    //element references
+    let $container;
+    let $buttonsContainer;
+    let $propertyListContaner;
+    let $searchInput;
+    let availableProperties = [];
+    let selectedProperties;
+    let search = '';
+
+    const parentGap = 20;
+
+    /**
+     * Lookup for characters in text to highlight
+     * @param {String} text - text to lookup
+     * @param {String} highlight - character(s) to be highlighted
+     * @param {regExp|String} match - match to be applied in the text
+     * @returns {String} - highlighted text
+     */
+    function highlightCharacter(text, highlight, match) {
+        return text.replace(match, `<b>${highlight}</b>`);
+    }
+
+    /**
+     * Creates property description list element
+     * @param {*} property
+     * @returns JQuery element containing property description
+     */
+    function createPropertyOption(property) {
+        const descriptionData = Object.assign({}, property);
+        if (search !== '') {
+            descriptionData.label = highlightCharacter(descriptionData.label, search, search);
+        }
+        const $propertyDescription = $(propertyDescriptionTpl({ property: descriptionData }));
+        const $checkboxContainer = $('.checkbox-container', $propertyDescription);
+        $checkboxContainer.append(checkBoxTpl({ id: descriptionData.id, checked: descriptionData.selected }));
+        const $checkbox = $('input', $checkboxContainer);
+        $checkbox.on('change', function () {
+            if (this.checked) {
+                selectedProperties.add(property.id);
+            } else {
+                selectedProperties.delete(property.id);
+            }
+        });
+        return $propertyDescription;
+    }
+
+    const instance = component({
+        positionContainer: function positionContainer() {
+            let { top, left, right, bottom } = this.config.data.position;
+            let maxHeight;
+            if (typeof bottom === 'undefined') {
+                maxHeight = $container.parent().height() - top - parentGap;
+            }
+            if (typeof top === 'undefined') {
+                maxHeight = $container.parent().height() - bottom - parentGap;
+            }
+            $container.css({ top, left, right, bottom, maxHeight });
+        },
+
+        /**
+         * Updates the list
+         */
+        redrawList: function redrawList() {
+            $propertyListContaner.empty();
+            availableProperties.forEach(property => {
+                property.selected = selectedProperties.has(property.id);
+                if (search === '' || property.label.toLowerCase().includes(search.toLowerCase())) {
+                    $propertyListContaner.append(createPropertyOption(property));
```

Comment:
**issue:** There is a performance impact adding DOM elements one by one, it will trigger a repaint so many times in a row. Please consider using a template for this, or apply a cache mechanism to perform the construction in memory only before injecting the HTML to the page.

Reply:
Thanks, implemented upon your suggestion